File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/AvailableArchiveWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import android.util.Base64
import com.google.gson.GsonBuilder
import org.xmtp.android.library.libxmtp.AvailableArchive

class AvailableArchiveWrapper {
    companion object {
        private val gson = GsonBuilder().create()

        fun encodeToObj(archive: AvailableArchive): Map<String, Any> = mapOf(
            "pin" to archive.pin,
            "metadata" to ArchiveMetadataWrapper.encode(archive.metadata),
            "sentByInstallation" to Base64.encodeToString(archive.sentByInstallation, Base64.NO_WRAP),
        )

        fun encodeList(archives: List<AvailableArchive>): String {
            val list = archives.map { encodeToObj(it) }
            return gson.toJson(list)
        }
    }
}

File: ios/Wrappers/AvailableArchiveWrapper.swift
import Foundation
import XMTP

struct AvailableArchiveWrapper {
	let pin: String

	init(_ archive: XMTP.AvailableArchive) {
		self.pin = archive.pin
	}

	func toJsonObject() -> [String: Any] {
		["pin": pin]
	}

	static func encodeList(_ archives: [XMTP.AvailableArchive]) throws -> String {
		let wrappers = archives.map { AvailableArchiveWrapper($0).toJsonObject() }
		let data = try JSONSerialization.data(withJSONObject: wrappers)
		guard let result = String(data: data, encoding: .utf8) else {
			throw WrapperError.encodeError("could not encode AvailableArchive list")
		}
		return result
	}
}
